### PR TITLE
[Flutter Parent][E2E][MBL-13953] | data seeding foundation + dashboard E2E test

### PR DESCRIPTION
**Pull request by joehoag**
This may look big, but of the 45 changed files, 24 are model/serialization-related.  So, still pretty big, but not humongous.

Some trickiness around data seeding and flutter driver testing:
(1) In order to get the auth code for a user, I had to call into native Android jsoup logic.  That's a little messy.
(2) In a flutter driver test, the *app* code (dashboard.dart) is free to call pretty much whatever it wants, but the *driver* code (i.e., the actual test logic, dashboard_test.dart) needs to avoid importing dart:ui.  A lot of production code pulls in dart:ui.  So: We seed data in app code, then use app<->driver communication to convey the seeding info to the driver and allow the driver to proceed.
(3) Routing to a particular screen is not easy in test driver code, because it doesn't have access to a context.  So I have to hold off the test driver code from starting the app until the seeding completes, at which point the app startup code should route to the dashboard screen.
(4) I can't access the L10n or AppLocalizations classes from flutter driver code (see (2)).  So I occasionally have to use raw strings as expected values.

Also, I'm open to the possibility of isolating all seeding models and apis to the test package.  My initial thought is that I need all of the seeding model classes to be referenced in serializers.dart anyway in order for them to be serialized properly, but if there is a way to bifurcate the serializers.dart file into a prod portion and a test portion I could work on that.

### CODE REVIEW

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1023675&projectid=20)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

File: apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt [hunk 1]
```diff
@@ -16,15 +16,75 @@
...
 
 class MainActivity : FlutterActivity() {
+    private val AUTHCODE_CHANNEL = "GET_AUTH_CODE"
+
     override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
         GeneratedPluginRegistrant.registerWith(flutterEngine)
         OldAppMigrations.init(flutterEngine, applicationContext)
+
+        // Implement getAuthCode / JSoup support via MethodChannel
+        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, AUTHCODE_CHANNEL).setMethodCallHandler { call, result ->
```
Comment (joehoag) [inline apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt]: This is the ugly Android code that calls into jsoup to help get an auth code.

File: apps/flutter_parent/lib/screens/courses/courses_screen.dart [hunk 2]
```diff
@@ -140,6 +140,7 @@ class _CoursesScreenState extends State<CoursesScreen> {
 
     return Text(
       text,
+      key: Key("${course.courseCode}_grade"),
```
Comment (joehoag) [inline apps/flutter_parent/lib/screens/courses/courses_screen.dart]: I threw in a few keys to allow me to access widgets easier from test driver code.

File: apps/flutter_parent/test_driver/dashboard_test.dart [hunk 3]
```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2020 - present Instructure, Inc.
```
Comment (joehoag) [inline apps/flutter_parent/test_driver/dashboard_test.dart]: This is the actual dashboard test code.

**Comment by CalvinKern** [comment]
This a little bit of a big ask, but would you mind going through and reformatting each file that you touched? It gets a little hairy sometimes when things aren't formatted properly, then we go into a file to make changes and there's a bunch of things moving around.

The rest of the team has auto format turned on so that anytime you edit a file it does this for us. If you search for "Format code on save" it will be under preferences -> Languages and frameworks -> Flutter -> Editor. You'll want that and "Organize imports on save". If you enable that, it should be as simple as opening all those files and running the project again.

**Comment by joehoag** [comment]
> would you mind going through and reformatting each file that you touched?

Sure, I'll look into that.

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

**Comment by joehoag** [comment]
OK, touched code has been reformatted.

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1023809&projectid=20)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

**Comment by joehoag** [comment]
Reformatting round 2 (80->120 max line length) completed.

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1023810&projectid=20)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

**Comment by Blackchai** [comment]
> OK, touched code has been reformatted.
> 
> (It appears that, in code that I modified as opposed to created, the reformatting touched a lot more than my code. Which would seem to imply that auto-formatting may not be universally used by other team members.)

You might also need to adjust your dart specific line wrap. Go to Preferences, search for Dart, Code Style -> Dart, and make sure you see "Hard Wrap at 120". I think it defaults to 80.

**Comment by CalvinKern** [comment]
>Which would seem to imply that auto-formatting may not be universally used by other team members.

It is now, it wasn't when we first started the project. We've just been slowly retouching things and committing the formats as we go. Thanks again for reformatting things!

File: apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt [hunk 4]
```diff
@@ -12,19 +12,79 @@
...
+                                .method(Connection.Method.GET)
+                                .data("client_id", clientId)
+                                .data("response_type", "code")
+                                .data("redirect_uri", redirectUrl)
+                                .execute()
+                        val loginForm = loginPageResponse.parse().select("form").first() as FormElement
+                        loginForm.getElementById("pseudonym_session_unique_id").`val`(login)
+                        loginForm.getElementById("pseudonym_session_password").`val`(password)
+                        val authFormResponse = loginForm.submit().cookies(loginPageResponse.cookies()).execute()
+                        //print("authFormResponse: ${authFormResponse.toString()}")
```
Comment (Blackchai) [inline apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt]: Do we want to keep these print/log statements?

File: apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt [hunk 5]
```diff
@@ -12,19 +12,79 @@
...
+
+                        return null
+                    }
+
+                    override fun onPostExecute(nullResult: Void?) : Unit {
+                        result.success(authCode)
+                        //return null
+                    }
+                }.execute(call)
+                //}
```
Comment (Blackchai) [inline apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt]: Do we need this comment?

File: apps/flutter_parent/lib/models/dataseeding/create_user_info.dart [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+part 'create_user_info.g.dart';
+
+abstract class CreateUserInfo implements Built<CreateUserInfo, CreateUserInfoBuilder> {
+  @BuiltValueSerializer(serializeNulls: true)
+  static Serializer<CreateUserInfo> get serializer => _$createUserInfoSerializer;
+
+  CreateUserInfo._();
+  factory CreateUserInfo([void Function(CreateUserInfoBuilder) updates]) = _$CreateUserInfo;
+
+//  CreateUserInfo({
```
Comment (Blackchai) [inline apps/flutter_parent/lib/models/dataseeding/create_user_info.dart]: Do we still need this?

File: apps/flutter_parent/lib/models/dataseeding/seeded_user.dart [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+abstract class SeededUser implements Built<SeededUser, SeededUserBuilder> {
+  @BuiltValueSerializer(serializeNulls: true)
+  static Serializer<SeededUser> get serializer => _$seededUserSerializer;
+
+  SeededUser._();
+  factory SeededUser([void Function(SeededUserBuilder) updates]) = _$SeededUser;
+
+//  SeededUser({
+//    this.id,
```
Comment (Blackchai) [inline apps/flutter_parent/lib/models/dataseeding/seeded_user.dart]: comment needed?

File: apps/flutter_parent/lib/models/dataseeding/pseudonym.dart [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+part 'pseudonym.g.dart';
+
+abstract class Pseudonym implements Built<Pseudonym, PseudonymBuilder> {
+  @BuiltValueSerializer(serializeNulls: true)
+  static Serializer<Pseudonym> get serializer => _$pseudonymSerializer;
+
+  Pseudonym._();
+  factory Pseudonym([void Function(PseudonymBuilder) updates]) = _$Pseudonym;
+
+//  Pseudonym({
```
Comment (Blackchai) [inline apps/flutter_parent/lib/models/dataseeding/pseudonym.dart]: comment needed?

File: apps/flutter_parent/lib/models/dataseeding/oauth_token.dart [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+part 'oauth_token.g.dart';
+
+abstract class OAuthToken implements Built<OAuthToken, OAuthTokenBuilder> {
+  @BuiltValueSerializer(serializeNulls: true)
+  static Serializer<OAuthToken> get serializer => _$oAuthTokenSerializer;
+
+  OAuthToken._();
+  factory OAuthToken([void Function(OAuthTokenBuilder) updates]) = _$OAuthToken;
+
+//  OAuthToken({
```
Comment (Blackchai) [inline apps/flutter_parent/lib/models/dataseeding/oauth_token.dart]: comment needed?

File: apps/flutter_parent/lib/screens/courses/courses_screen.dart [hunk 10]
```diff
@@ -109,9 +109,11 @@ class _CoursesScreenState extends State<CoursesScreen> {
               crossAxisAlignment: CrossAxisAlignment.start,
               children: <Widget>[
                 SizedBox(height: 8),
-                Text(course.name ?? '', style: Theme.of(context).textTheme.subhead),
+                Text(course.name ?? '',
```
Comment (Blackchai) [inline apps/flutter_parent/lib/screens/courses/courses_screen.dart]: I think these lines here were probs part of the formatting prior to line wrap being at 120, they look fine for me locally.

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

File: apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt (hunk 4 above)
Comment (joehoag) [reply apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt]: Removed.

File: apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt (hunk 5 above)
Comment (joehoag) [reply apps/flutter_parent/android/app/src/main/kotlin/com/instructure/parentapp/MainActivity.kt]: Removed.

File: apps/flutter_parent/lib/models/dataseeding/create_user_info.dart (hunk 6 above)
Comment (joehoag) [reply apps/flutter_parent/lib/models/dataseeding/create_user_info.dart]: Removed.

File: apps/flutter_parent/lib/models/dataseeding/seeded_user.dart (hunk 7 above)
Comment (joehoag) [reply apps/flutter_parent/lib/models/dataseeding/seeded_user.dart]: Removed.

File: apps/flutter_parent/lib/models/dataseeding/pseudonym.dart (hunk 8 above)
Comment (joehoag) [reply apps/flutter_parent/lib/models/dataseeding/pseudonym.dart]: Removed.

File: apps/flutter_parent/lib/models/dataseeding/oauth_token.dart (hunk 9 above)
Comment (joehoag) [reply apps/flutter_parent/lib/models/dataseeding/oauth_token.dart]: Removed.

**Comment by joehoag** [comment]
Pushed an update that got rid of a lot of commented-out code.

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1024296&projectid=20)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

**Pull request merged**